R/checkModels.R

```diff
@@ -1,74 +1,201 @@
+    person.groups$group <- all.persons.lab
+    all.persons <- FALSE
+  }
+  # person.groups auf Plausibilitaet checken
+  if(!is.null(person.groups)){
+    # hat erste Spalte mehr Elemente als alle anderen
+    len <- sapply(person.groups, function(x) length(unique(x)))
+    len.log <- len < len[1]
+    if(!all(len.log[-1])) warning(paste0("first column of person.groups might not contain person ids; please check\n(number of unique elements is smaller than in another column)"),
+                                  call. = FALSE)
+  }
+
+  return(d)
+}
+
+### called by defineModel() and splitModels() ----------------------------------
+
+checkQmatrixConsistency <- function(qmat){
+  qmat <- eatTools::makeDataFrame(qmat, name = "Q matrix")
+  if(!inherits(qmat[,1], "character")){
+    qmat[,1] <- as.character(qmat[,1])}
+
+  nClass<- sapply(qmat[,-1,drop=FALSE], inherits, what=c("numeric", "integer"))
+  if(!all(nClass)){
+    warning(paste0("Found non-numeric indicator column(s) in the Q matrix. Transform column(s) '",
+                   paste(colnames(qmat)[ which(nClass==FALSE)+1], collapse = "', '") ,"' into numeric format."))
+    qmat <- data.frame(qmat[,1,drop=FALSE], eatTools::asNumericIfPossible(qmat[,-1,drop=FALSE]), stringsAsFactors = FALSE)}
+
+  werte <- eatTools::tableUnlist(qmat[,-1,drop=FALSE], useNA="always")
+  if(length(setdiff(names(werte), c("0","1", "NA")))<0){
+    stop("Q matrix must not contain entries except '0' and '1'.\n")}
+  if(werte[match("NA", names(werte))] > 0){
+    stop("Missing values in Q matrix.\n")}
+
+  wertes <- lapply(qmat[,-1,drop=FALSE], FUN = function(col) {all(col == 0)})
+  konst <- which(wertes == TRUE)
+  if(length(konst)>0){
+    cat(paste0("Column(s) '",paste(names(konst), collapse = "', '"),"' in Q matrix are constant with value 0. Delete column(s).\n"))
+    qmat <- qmat[,-match(names(konst), colnames(qmat)), drop=FALSE]
+  }
+
+  doppel <- which(duplicated(qmat[,1]))
+  if(length(doppel)>0){
+    cat("Found duplicated elements in the item id column of the q matrix. Duplicated elements will be removed.\n")
+    chk  <- table(qmat[,1])
+    chk  <- chk[which(chk > 1)]
+    chkL <- lapply(names(chk), FUN = function(ch){
+      qChk <- qmat[which(qmat[,1] == ch),]
+      pste <- apply(qChk, 1, FUN = function(x) {paste(x[-1], collapse="")})
+      if(!all(pste == pste[1])){
+        stop("Inconsistent q matrix.\n")}
+    })
+    qmat <- qmat[!duplicated(qmat[,1]),]
+  }
+
+  zeilen <- apply(qmat, 1, FUN = function(y) {all(names(table(y[-1])) == "0")} )
+  weg    <- which(zeilen == TRUE)
+  if(length(weg)>0) {
+    cat(paste("Note: Following ",length(weg)," item(s) in Q matrix do not belong to any dimension. Delete these item(s) from Q matrix.\n",sep=""))
+    cat("    "); cat(paste(qmat[weg,1],collapse=", ")); cat("\n")
+    qmat <- qmat[-weg,]
+  }
+
+  # wenn nur eine Spalte wird diese als IDs angenommen
```

Comment:
These checks were in `splitModels`. `defineModel` also uses a qMatrix. Is it the same kind of qMatrix? 

Because in the old pull request I deleted these lines:
![grafik](https://github.com/user-attachments/assets/e4c5feb9-6338-4e95-b6a0-6609af64b996)

and added these to the checkQMatrix/checkPersonGroups.. functions, as you can see here and line 23f:
![grafik](https://github.com/user-attachments/assets/d92a64e5-93a1-4725-8cd3-945a5a75379b)

Should I leave the function as is then, and delete these lines from `splitModels`? 
I didn't run the examples yet.

Reply:
Yes, it is the same kind of Q matrix. As the functions in `defineModel()` and `splitModels()` were developed independently from each other (and by different persons), programming is a little redundant. The function `checkQmatrixConsistency()` was added to the package at a later date. I would suggest that in `splitModels()` you replace the passage you marked with "what to do with this" simply with `checkQmatrixConsistency()`. However, I would not like to do this “blindly”, but would rather go through it line by line to see whether what is checked in `splitModels()` is already implemented by `checkQmatrixConsistency()` (and vice versa). That would be a somewhat lengthy and cumbersome procedure. Should I do that or would you like to?